Test and refactor Homestay model

rbmrclo opened the pull request:
TODO:
- [x] Test `Homestay` model
- [x] Refactor `Homestay` model

Changelog:
- Deleted unused code in homestay model https://github.com/pethomestay/phs/commit/beaa7dce73c9cb351ae06ae4f8708a5b6c5c1805
- Homestay I18n support https://github.com/pethomestay/phs/commit/cf0d0d6fe90fa98a6de7f01874b08f9eeff5d91b, https://github.com/pethomestay/phs/commit/74f6369cdf8c3aefb5eda337996f664eb325baa3
- Homestay factory https://github.com/pethomestay/phs/commit/462ac2af21885bc0c6d35232e9ed047dfedb5e93, https://github.com/pethomestay/phs/commit/53adacfc0bd2f31d06bccef51550ca295825160b,
  https://github.com/pethomestay/phs/commit/385391cc96da5b1545e60f68805f76c1f1e4b91d
- Homestay decorator https://github.com/pethomestay/phs/commit/292a8f54869fae714dc2b0398e36144305cba3b0, https://github.com/pethomestay/phs/commit/cd5d0ae514445ef7b2d5858786cee56238b32021, https://github.com/pethomestay/phs/commit/ebeb194189d1813f710b969179c5a344407c3c77
- Homestay model https://github.com/pethomestay/phs/commit/237e273614c0f7209f5639cab2d68e5ab8f0c996, https://github.com/pethomestay/phs/commit/e91c1802acfb42844ef75d6ac3df710ea0438a83, https://github.com/pethomestay/phs/commit/031afb6347910b53864b0c748f7dca381798ec74, https://github.com/pethomestay/phs/commit/5edc865484acfe57ed2da3337830cb44af5fa137,
  https://github.com/pethomestay/phs/commit/920aeab89175095532471994680b443410aa3556
- Specs for homestay model callbacks https://github.com/pethomestay/phs/commit/52b20d4bf0760dbe5a024fd24f7f8e71facf0bf3

rbmrclo committed:
Adds homestay factory

rbmrclo committed:
Adds homestay decorator + extract methods from homestay model

rbmrclo committed:
Adds initial specs for homestay model

- Associations
- Serializations
- Validations
- Nested attributes acceptance
- Delegations

rbmrclo committed:
Updates homestay model

- Remove extracted methods by Homestay decorator
- Implement delegations

rbmrclo committed:
Updates user factory

rbmrclo committed:
Improves accuracy of specs for numericality validation of homestay model

rbmrclo committed:
Updates homestay factory

- Title & slug should be in sequence

rbmrclo committed:
Refactor homestay model

- Append “?” for boolean values & rename methods for intuitiveness
- Adds comments for deprecated/unused code

rbmrclo committed:
Adds specs for scopes and instance methods in homestay model

rbmrclo committed:
Append question mark to boolean values for intuitiveness

adimasuhid commented on app/decorators/homestay_decorator.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def auto_decline_sms_text
+    auto_decline_sms.present? ? auto_decline_sms : auto_decline_sms_template
```
If `auto_decline_sms` returns nil, this might be more elegant: 

```
auto_decline_sms || auto_decline_sms_template
```

rbmrclo replied:
The reason why I preferred to use `.present?` versus the [previous implementation](https://github.com/pethomestay/phs/pull/11/files#diff-4e45e37997862be1fd1c53582aa39f97L115) is because `auto_decline_sms` is a [text](https://github.com/pethomestay/phs/blob/master/db/schema.rb#L222). It's possible that it could have empty `""` string values given that it's not being required during validation phase of object lifecycle.

Take note that:

``` ruby
irb> "" || "hello"
irb> ""
```

I agree that the above mentioned code is more elegant, but i'm sacrificing that over data-error invulnerability. What do you think? @adimasuhid

adimasuhid replied:
Yeah good point. Was thinking about that as well earlier, this should be fine then.

adimasuhid commented on app/decorators/homestay_decorator.rb:
```diff
@@ -0,0 +1,64 @@
+  end
+
+  def auto_interest_sms_text
+    auto_interest_sms.present? ? auto_interest_sms : auto_interest_sms_template
```
Same as above

adimasuhid commented on app/decorators/homestay_decorator.rb:
```diff
@@ -0,0 +1,64 @@
+      "Sorry - I can't help this time, but please ask again in the future!"
+    end
+
+end
```
Most of the method returns here are string messages. I wonder if this would make sense in a yaml file? Kind of like en.yml

rbmrclo replied:
I agree with this. Normally, it's ideal to see this kind of code being localized. I'm wondering if we have plans to support `I18n`? If yes, i'd be happy to apply it. cc @zero51

zero51 replied:
+1 from me. Even if there aren't plans to support additional locales, this has no place in the model. Would be more than happy to move them to the default en.yml locale file.

adimasuhid commented on app/models/homestay.rb:
```diff
@@ -34,43 +35,53 @@ class Homestay < ActiveRecord::Base
+
+  validates :delivery_radius, numericality: {
     greater_than_or_equal_to: 0, only_integer: true
   }, if: 'delivery_price.present?'
```
This might be a rails 4 implementation, though this might be more elegant:

```
if: -> { delivery_price.present? }
```

Also the with the number of validations, would it make sense to add a HomestayValidator? Interested in your thoughts @rbmrclo @zero51

adimasuhid commented on app/models/homestay.rb:
```diff
@@ -250,10 +223,14 @@ def favourite?(current_user)
   end
 
+  def decorator
+    HomestayDecorator.new(self)
```
Interesting delegation. I like it!

rbmrclo committed:
Adds specs for HomestayDecorator#pretty_services and Homestay#energy_levels

rbmrclo committed:
Moves texts from decorators to config/locales/en.yml

rbmrclo committed:
Move Homestay#pretty_services to decorator + i18n support

- Adds code comment too for ReferenceData::EnergyLevel

zero51 commented:
Looks good to me.

rbmrclo committed:
Write specs for homestay model callbacks

rbmrclo committed:
Deleted unused code in homestay model

Class Methods:
- .reject_unavailable_homestays
- .homestay_ids_unavailable_between
- .available_between
- .homestay_ids_booked_between
- .not_booked_between

Instance Methods:
- #supervision?

rbmrclo commented:
@zero51 @adimasuhid This one is cooked and ready to be served, please review! Thanks!

zero51 commented:
All looks good to me!

rbmrclo commented:
Moved to https://github.com/pethomestay/phs/pull/21
